rfcs/0132-meson-builds-nix.md

```diff
@@ -0,0 +1,223 @@
+
+## Learning the new system
+
+A somewhat subjective but non-negligible issue is the entry barrier of this new build system. Switching from a known build system to one unknown is not without its problems.
+
+However, the Meson development team strives to keep the DSL easy to learn and pleasurable to use. It should not be hard to become familiar with the Python-esque syntax of meson, and its functional, stateless approach is certainly a feature highly appreciated by the Nix community as a whole.
+
+The huge advantages of implementing Meson surpass the small drawbacks of learning it.
+
+## Source code changes
+
+Further to the obvious inclusion of meson files (and the removal of the old quasi-autotools ones), there is a reasonable expectation of code refactoring.
+
+However, such refactorings are completely validated on the long term goals of Nix, in particular the improvements on portability.
+
+## End users
+
+The most known end user of Nix is certainly Nixpkgs. However, there are many other Linux distributions that already keep Nix on their repositories (15 families, according to Repology). There is also a reasonable expectation of affecting those package managers' devteams.
+
+However, most (if not all) of those distributions already have Meson and its companion tool Ninja in their respective package databases (53 families, according to Repology), given that many open source projects use them as build system.
+
+## Transition between old and new build infrastructure
+
+The transition between old and new build systems should be as smooth and controlled as possible.
+
+# Alternatives
+[alternatives]: #alternatives
+
+The alternatives are
+
+- Do nothing
+
+  It would keep the current code confusing and harder to work with, as stated on the motivation section.
+
+- Use CMake
+
+  Indeed, CMake has many noteworthy advantages:
+
+  - Supports typical high level idiomatic constructions.
+  - Can generate GMake- and Ninja-compatible scripts.
+    - By design, Meson does not provide a Make backend.
+  - Both Meson and CMake support Microsoft Windows NT and Apple MacOS platforms
+    - As well as MSVC and XCode programming environments.
+
+  However, CMake DSL is arguably more complex and cumbersome, whereas Meson is more polished.
+
+- Evaluate other building systems (waf, premake, bazel, xmake etc.)
+
+  About this, a principle should be observed:
+
+  Per the bootstrap route issue discussed above, build tools strongly tied to other programming languages are severely discouraged.
+
+  E.g. waf is basically a Python library, whereas premake and xmake are Lua libraries. They can't be decoupled of their implementation languages.
+
+  Another indispensable attribute to be considered is cross-compilation support.
+
+- Use Bazel
+
+  For the sake of completeness, there is Bazel, a Google(TM)-backed build system that sells itself as "fast, scalable, multi-language and extensible".
```

Comment:
Just as a comment, adding (TM) after "Google" seems a bit snarky/edgy/weird especially considering that "Meson is a registered trademark of Jussi Pakkanen" per https://mesonbuild.com/legal.html - so should every mention of it in this RFC also be rewritten to "Meson(TM)"?

I get that apparently Bazel or Google isn't popular or it might also not be the right choice for building/bootstrapping `nix`, but this whole section feels very unprofessional in style and argumentation, both in the document itself as well as the discussion here in the PR.

If you for example just want a simple way to define build steps and save time on rebuilds, you could even just use https://gittup.org/tup/ instead of adopting a complicated system like Meson.

Reply:
I have years experience with tup. It is a really bad choice, because tup projects lack modular structure.

The only alternative to c and c++ projects is CMake.

Reply:
> Just as a comment, adding (TM) after "Google" seems a bit snarky/edgy/weird especially considering that "Meson is a registered trademark of Jussi Pakkanen" per https://mesonbuild.com/legal.html - so should every mention of it in this RFC also be rewritten to "Meson(TM)"?

Another possible reason for the snark is because *specifically* Google is well known for:
- a vast graveyard of abandoned projects
- designing the software they don't abandon for the specific use case of working well in Google's own heavily specific environment, many of which boil down to "bazel wants to use its own internal vendored everything, because at google, they have one official everything baked into bazel and then make all software use that via bazel".

Reply:
I don't see how this explains this sort of wording or style of discussion/argumentation. If Bazel is a bad fit, I'm sure this can be phrased in a better way with clearer arguments than "Backed by Google(TM)."

It might be easier if there are clear goals and clear no-gos stated in the RFC (e.g. "bootstrapping `nix` with its build system SHOULD not require significantly more dependencies than bootstrapping all build tools and building the `nix` binaries itself" - which would exclude Bazel, since it requires a JVM, or the Python implementation of Meson, since it requires a Python interpreter). Ideally even with performance numbers that show that there is actually an improvement in the stated expected goals (which don't include "keeping the bootstrap dependency graph small" btw. even though a large portion of the text of the RFC and the discussions seem to center around this).

Reply:
You need/want a less sarcastic paragraph? I can possibly reword it.

On the other hand, my intention with this RFC is not to advocate generically for some better build system.
Rather, it is to advocate specificallty for Meson.

Reply:
I would also like to point out that glibc depends on python to build itself for a while now. Python on its own is not actually a meaningful additional dependency *unless* you are committed to equally supporting musl, which my understanding is that NixOS isn't, even if `nix` is.

Reply:
Yet another good reason to push for Muon+Samurai.